### PR TITLE
Photon vision

### LINKED ISSUE #5: Add Photon Vision Integration

### PR DESCRIPTION
**Pull request by agrinmanriv0537**
Closes #5 - Photon Vision Integration

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/swerve/SwerveReal.java [hunk 1]
```diff
@@ -11,14 +15,41 @@ public class SwerveReal implements SwerveIO {
...
+            inputs.results[i] = cameras[i].photonCamera.getLatestResult();
+            inputs.seesTarget[i] = cameras[i].seesTarget();
+            inputs.estimatedRobotPose[i] = cameras[i].getEstimatedGlobalPose(previousPose);
+        }
+
+    }
+
+    public Optional<EstimatedRobotPose> getEstimatedGlobalPose(int i,
+        Pose2d prevEstimatedRobotPose) {
+        return this.cameras[i].getEstimatedGlobalPose(prevEstimatedRobotPose);
```
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/swerve/SwerveReal.java]: ur not actually using this function

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java [hunk 2]
```diff
@@ -22,18 +25,27 @@
...
+    // .withProperties(Map.of("Color when true", "green", "Color when false", "red"))
+    // .withPosition(8, 4).withSize(2, 2).getEntry();
 
     /**
      * Swerve Subsystem
      */
     public Swerve(SwerveIO swerveIO) {
         this.swerveIO = swerveIO;
-        swerveIO.updateInputs(inputs);
+        swerveIO.updateInputs(inputs, swerveOdometry.getEstimatedPosition());
```
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/swerve/Swerve.java:48]: this needs to be after line 64. `swerveOdometry` is null until then.

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java [hunk 3]
```diff
@@ -202,8 +222,55 @@ public void resetModulesToAbsolute() {
     @Override
     public void periodic() {
         swerveOdometry.update(getGyroYaw(), getModulePositions());
-        swerveIO.updateInputs(inputs);
+        swerveIO.updateInputs(inputs, swerveOdometry.getEstimatedPosition());
         Logger.processInputs("Swerve", inputs);
+        for (double latency : inputs.latencies) {
+            latencyGood = latency < 0.6 ? true : false;
```
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/swerve/Swerve.java:228]: this is whatever is last, so always the value of camera 4

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java [hunk 4]
```diff
@@ -202,8 +222,55 @@ public void resetModulesToAbsolute() {
...
+        }
+        SmartDashboard.putBoolean("photonGood", latencyGood);
+        Rotation2d yaw = Rotation2d.fromDegrees(inputs.yaw);
+        swerveOdometry.update(yaw, getSwerveModulePositions());
+        if (!hasInitialized /* || DriverStation.isDisabled() */) {
+            var robotPose = inputs.positions[0];
```
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/swerve/Swerve.java:234]: we'll have to be careful with this.  We need to make sure this camera is always in view of a tag no matter the starting position.

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java [hunk 5]
```diff
@@ -202,8 +222,55 @@ public void resetModulesToAbsolute() {
...
+            }
+        } else {
+            for (Optional<EstimatedRobotPose> estimatedPose : inputs.estimatedRobotPose) {
+                if (estimatedPose.isPresent()) {
+                    var camPose = estimatedPose.get();
+                    if (camPose.targetsUsed.get(0).getArea() > 0.7) {
+                        swerveOdometry.addVisionMeasurement(camPose.estimatedPose.toPose2d(),
+                            camPose.timestampSeconds);
+                    }
+                    field.getObject("Cam Est Pose").setPose(camPose.estimatedPose.toPose2d());
```
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/swerve/Swerve.java:248]: Do u want 1 value or 1 per camera???

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java [hunk 6]
```diff
@@ -250,4 +325,5 @@ public static boolean shouldFlipPath() {
         }
         return false;
     }
+
```
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/swerve/Swerve.java:328]: no

File: src/main/java/frc/robot/subsystems/swerve/SwerveReal.java (hunk 1 above)
Comment (agrinmanriv0537) [reply src/main/java/frc/robot/subsystems/swerve/SwerveReal.java]: function deleted.

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java (hunk 2 above)
Comment (agrinmanriv0537) [reply src/main/java/frc/robot/subsystems/swerve/Swerve.java:52]: moved.

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java (hunk 3 above)
Comment (agrinmanriv0537) [reply src/main/java/frc/robot/subsystems/swerve/Swerve.java]: fixed

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java (hunk 4 above)
Comment (agrinmanriv0537) [reply src/main/java/frc/robot/subsystems/swerve/Swerve.java:238]: yas

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java (hunk 5 above)
Comment (agrinmanriv0537) [reply src/main/java/frc/robot/subsystems/swerve/Swerve.java:258]: 4 cameras fixed

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java (hunk 6 above)
Comment (agrinmanriv0537) [reply src/main/java/frc/robot/subsystems/swerve/Swerve.java:338]: maybe

**Review by legoguy1000 — APPROVED** [review]

**Pull request merged**